src/ui_fsmenu/addons/progress.cc

```diff
@@ -80,4 +81,121 @@ ProgressIndicatorWindow::ProgressIndicatorWindow(UI::Panel* parent,
 	initialization_complete();
 }
 
+struct HangupWindow : public UI::Window {
+	static std::map<uint32_t, HangupWindow*> windows;
+
+	HangupWindow(UI::Panel& parent,
+	             UI::WindowStyle style,
+	             AddOns::NetAddons& net_addons,
+	             uint32_t id)
+	   : UI::Window(&parent,
+	                style,
+	                "hangup",
+	                0,
+	                0,
+	                parent.get_inner_w() / 2,
+	                2 * kRowButtonSize,
+	                _("Server Unresponsive")),
+	     id_(id),
+	     net_addons_(net_addons),
+	     modal_(*this),
+	     box_(this, panel_style_, "main_box", 0, 0, UI::Box::Vertical, get_inner_w()),
+	     txt1_(&box_,
+	           panel_style_,
+	           "text",
+	           style == UI::WindowStyle::kFsMenu ? UI::FontStyle::kFsMenuInfoPanelParagraph :
+	                                               UI::FontStyle::kWuiInfoPanelParagraph,
+	           _("The server is not responding…"),
+	           UI::Align::kCenter),
+	     txt2_(&box_,
+	           panel_style_,
+	           "time",
+	           style == UI::WindowStyle::kFsMenu ? UI::FontStyle::kFsMenuInfoPanelParagraph :
+	                                               UI::FontStyle::kWuiInfoPanelParagraph,
+	           "",
+	           UI::Align::kCenter),
+	     interrupt_(&box_,
+	                "interrupt",
+	                0,
+	                0,
+	                0,
+	                0,
+	                style == UI::WindowStyle::kFsMenu ? UI::ButtonStyle::kFsMenuSecondary :
+	                                                    UI::ButtonStyle::kWuiSecondary,
+	                _("Cancel")) {
+		box_.add(&txt1_, UI::Box::Resizing::kExpandBoth);
+		box_.add_space(kRowButtonSpacing);
+		box_.add(&txt2_, UI::Box::Resizing::kExpandBoth);
+		box_.add_space(kRowButtonSpacing);
+		box_.add(&interrupt_, UI::Box::Resizing::kAlign, UI::Align::kCenter);
+
+		interrupt_.sigclicked.connect([this, &net_addons]() {
+			delete this;
+			net_addons.interrupt();
+		});
+
+		set_center_panel(&box_);
+		center_to_parent();
+
+		windows.emplace(id_, this);
+		initialization_complete();
+	}
+
+	~HangupWindow() override {
+		windows.erase(id_);
+	}
+
+	void set_elapsed(uint32_t elapsed) {
+		txt2_.set_text(gametimestring(elapsed, true));
+		do_redraw_now();
+	}
+
+	void die() override {
+		AddOns::NetAddons& net_addons = net_addons_;
+		delete this;
+		net_addons.interrupt();
+	}
+
+private:
+	uint32_t id_;
+	AddOns::NetAddons& net_addons_;
+	UI::Panel::ModalGuard modal_;
+	UI::Box box_;
+	UI::Textarea txt1_, txt2_;
+	UI::Button interrupt_;
+};
+
+std::map<uint32_t, HangupWindow*> HangupWindow::windows;
```

Comment:
this seeems to be a declaration that should have been made in the header. 
Generally I have the feelinmg that declaration and implementation are not very cleanly seperated in this PR, which m,akes issue even harder to understand for me.